File: ios/Flashmart/Data/OnboardingDocumentStorage.swift
import Foundation

struct OnboardingDocumentSlot: Equatable, Identifiable {
    var id: String { label }
    let label: String
    let required: Bool
}

struct PendingOnboardingDocument: Equatable, Identifiable {
    var id: String { label }
    let label: String
    let displayName: String
    let localPath: URL
    let mimeType: String?
    let fileSize: Int64

    var localURL: URL { localPath }
}

enum OnboardingDocumentStorage {
    static func dealerSlots() -> [OnboardingDocumentSlot] {
        withDocumentPolicy(dealerDocumentSlots())
    }

    static func shopkeeperSlots() -> [OnboardingDocumentSlot] {
        withDocumentPolicy(shopkeeperDocumentSlots())
    }

    static func validateRequired(
        slots: [OnboardingDocumentSlot],
        documents: [PendingOnboardingDocument]
    ) -> String? {
        guard AppConfig.requireOnboardingDocuments else { return nil }
        return validateRequiredDocuments(slots: slots, documents: documents)
    }

    private static func withDocumentPolicy(_ slots: [OnboardingDocumentSlot]) -> [OnboardingDocumentSlot] {
        guard AppConfig.requireOnboardingDocuments else {
            return slots.map { OnboardingDocumentSlot(label: $0.label, required: false) }
        }
        return slots
    }

    static func dealerDocumentSlots() -> [OnboardingDocumentSlot] {
        [
            OnboardingDocumentSlot(label: "ID proof (Aadhaar / PAN)", required: true),
            OnboardingDocumentSlot(label: "GST certificate", required: true),
            OnboardingDocumentSlot(label: "Business registration / trade license", required: true),
            OnboardingDocumentSlot(label: "Bank proof / cancelled cheque", required: false),
        ]
    }

    static func shopkeeperDocumentSlots() -> [OnboardingDocumentSlot] {
        [
            OnboardingDocumentSlot(label: "ID proof (Aadhaar / PAN)", required: true),
            OnboardingDocumentSlot(label: "Shop license / rent agreement", required: true),
            OnboardingDocumentSlot(label: "GST certificate", required: false),
        ]
    }

    static func validateRequiredDocuments(
        slots: [OnboardingDocumentSlot],
        documents: [PendingOnboardingDocument]
    ) -> String? {
        let labels = Set(documents.map(\.label))
        let missing = slots.filter { $0.required && !labels.contains($0.label) }.map(\.label)
        guard !missing.isEmpty else { return nil }
        return "Please upload: \(missing.joined(separator: ", "))"
    }

    static func stageDocument(from sourceURL: URL, label: String) -> PendingOnboardingDocument? {
        let displayName = sourceURL.lastPathComponent.isEmpty ? "document" : sourceURL.lastPathComponent
        let stagingDir = onboardingRoot.appendingPathComponent("staging", isDirectory: true)
        try? FileManager.default.createDirectory(at: stagingDir, withIntermediateDirectories: true)
        let ext = sourceURL.pathExtension.isEmpty ? "bin" : sourceURL.pathExtension
        let target = stagingDir.appendingPathComponent("\(UUID().uuidString).\(ext)")
        do {
            if FileManager.default.fileExists(atPath: target.path) {
                try FileManager.default.removeItem(at: target)
            }
            try FileManager.default.copyItem(at: sourceURL, to: target)
            let attrs = try FileManager.default.attributesOfItem(atPath: target.path)
            let size = (attrs[.size] as? NSNumber)?.int64Value ?? 0
            return PendingOnboardingDocument(
                label: label,
                displayName: displayName,
                localPath: target,
                mimeType: mimeType(for: ext),
                fileSize: size
            )
        } catch {
            return nil
        }
    }

    static func persistForUser(userId: String, pending: [PendingOnboardingDocument]) -> [OnboardingDocument] {
        let userDir = onboardingRoot.appendingPathComponent(userId, isDirectory: true)
        try? FileManager.default.createDirectory(at: userDir, withIntermediateDirectories: true)
        return pending.map { doc in
            let safeName = doc.displayName.replacingOccurrences(of: #"[^\w.\-]+"#, with: "_", options: .regularExpression)
            let target = userDir.appendingPathComponent(safeName)
            try? FileManager.default.removeItem(at: target)
            try? FileManager.default.copyItem(at: doc.localPath, to: target)
            try? FileManager.default.removeItem(at: doc.localPath)
            return OnboardingDocument(
                id: "doc-\(UUID().uuidString)",
                label: doc.label,
                fileName: doc.displayName,
                mimeType: doc.mimeType,
                fileSize: doc.fileSize,
                uploadedAt: ISO8601DateFormatter().string(from: Date())
            )
        }
    }

    private static var onboardingRoot: URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        return base.appendingPathComponent("onboarding", isDirectory: true)
    }

    private static func mimeType(for ext: String) -> String {
        switch ext.lowercased() {
        case "pdf": return "application/pdf"
        case "jpg", "jpeg": return "image/jpeg"
        case "png": return "image/png"
        default: return "application/octet-stream"
        }
    }
}

File: android/app/src/main/java/com/mart/distribution/demo/data/onboarding/OnboardingDocumentStorage.kt
package com.mart.distribution.demo.data.onboarding

import android.content.Context
import android.net.Uri
import android.provider.OpenableColumns
import android.webkit.MimeTypeMap
import com.mart.distribution.demo.BuildConfig
import com.mart.distribution.demo.data.api.dto.OnboardingDocumentDto
import java.io.File
import java.util.UUID

data class PendingOnboardingDocument(
    val label: String,
    val displayName: String,
    val localPath: String,
    val mimeType: String?,
    val fileSize: Long,
)

object OnboardingDocumentStorage {
    fun dealerDocumentSlots(): List<OnboardingDocumentSlot> =
        withDocumentPolicy(
            listOf(
                OnboardingDocumentSlot("ID proof (Aadhaar / PAN)", required = true),
                OnboardingDocumentSlot("GST certificate", required = true),
                OnboardingDocumentSlot("Business registration / trade license", required = true),
                OnboardingDocumentSlot("Bank proof / cancelled cheque", required = false),
            ),
        )

    fun shopkeeperDocumentSlots(): List<OnboardingDocumentSlot> =
        withDocumentPolicy(
            listOf(
                OnboardingDocumentSlot("ID proof (Aadhaar / PAN)", required = true),
                OnboardingDocumentSlot("Shop license / rent agreement", required = true),
                OnboardingDocumentSlot("GST certificate", required = false),
            ),
        )

    private fun withDocumentPolicy(slots: List<OnboardingDocumentSlot>): List<OnboardingDocumentSlot> =
        if (BuildConfig.REQUIRE_ONBOARDING_DOCUMENTS) {
            slots
        } else {
            slots.map { slot -> slot.copy(required = false) }
        }

    fun stageDocument(
        context: Context,
        uri: Uri,
        label: String,
    ): PendingOnboardingDocument? {
        val resolver = context.contentResolver
        val displayName =
            resolver.query(uri, arrayOf(OpenableColumns.DISPLAY_NAME, OpenableColumns.SIZE), null, null, null)
                ?.use { cursor ->
                    if (cursor.moveToFirst()) {
                        val nameIdx = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                        if (nameIdx >= 0) cursor.getString(nameIdx) else null
                    } else {
                        null
                    }
                }
                ?: uri.lastPathSegment
                ?: "document"
        val mimeType = resolver.getType(uri)
        val stagingDir = File(context.filesDir, "onboarding/staging").apply { mkdirs() }
        val ext =
            MimeTypeMap.getSingleton()
                .getExtensionFromMimeType(mimeType ?: "")
                ?.takeIf { it.isNotBlank() }
                ?: displayName.substringAfterLast('.', "bin")
        val outFile = File(stagingDir, "${UUID.randomUUID()}.$ext")
        resolver.openInputStream(uri)?.use { input ->
            outFile.outputStream().use { output -> input.copyTo(output) }
        } ?: return null
        return PendingOnboardingDocument(
            label = label,
            displayName = displayName,
            localPath = outFile.absolutePath,
            mimeType = mimeType,
            fileSize = outFile.length(),
        )
    }

    fun persistForUser(
        context: Context,
        userId: String,
        pending: List<PendingOnboardingDocument>,
    ): List<OnboardingDocumentDto> {
        val userDir = File(context.filesDir, "onboarding/$userId").apply { mkdirs() }
        return pending.map { doc ->
            val target = File(userDir, doc.displayName.replace(Regex("[^\\w.\\-]+"), "_"))
            File(doc.localPath).copyTo(target, overwrite = true)
            File(doc.localPath).delete()
            OnboardingDocumentDto(
                id = "doc-${UUID.randomUUID()}",
                label = doc.label,
                fileName = doc.displayName,
                mimeType = doc.mimeType,
                fileSize = doc.fileSize,
                uploadedAt = java.time.Instant.now().toString(),
                localPath = target.absolutePath,
            )
        }
    }

    fun fileForDocument(document: OnboardingDocumentDto): File? =
        document.localPath?.let { path ->
            val file = File(path)
            if (file.exists()) file else null
        }
}

data class OnboardingDocumentSlot(
    val label: String,
    val required: Boolean,
)
